### PR TITLE
Handle migrated celo headers

### PR DESCRIPTION
**Pull request by piersy**
This PR updates the header encoding and decoding for celo blocks to be able to support data migrated from celo-blockchain.

Our post gingerbread headers actually decode into the op-geth header structure without any changes. Since adding fields to the end of a struct does not cause rlp decoding to fail for previously encoded values. Our pre-gingerbread don't decode into the op-geth header structure because the pre-gingerbread headers miss fields from the middle of the struct so we need to handle those explicitly.

The decision on how to decode is made based on introspecting the rlp and looking at the second field which is a hash (length 32 bytes) post-gingerbread, but an address (length 20 bytes) pre-gingerbread.

The decision on how to encode is made based on the value of gasLimit if zero then it's a pre-gingerbread header.

Additionally we needed to allow genesis blocks in a CeL2 migrated chain context (by checking if `Cel2Time` is set and non zero) that lack GasLimit and Difficulty because those fields were considered required by op-geth but not present in early Celo.

Finally some tests that used blocks with zero gas limits needed to be updated since with a zero gas limit the different encoding was now breaking them, and adding a non zero gas limit meant the block hash changed.

### CODE REVIEW

**Comment by palango** [comment]
> The decision on how to encode is made based on the value of gasLimit if zero then it's a pre-gingerbread header.

I thought we would set that to the correct value, so that the apis return something useful. Or is this done somewhere in the API layer?

> Additionally we needed to allow genesis blocks in a CeL2 context (by checking if Cel2Time is set) that lack GasLimit and Difficulty because those fields were considered required by op-geth but not present in early Celo.

Is this requires so that the genesis hash stays the same, or why can't we just add those values to the celo genesis block?

> Many tests using TestChainConfig failed because it set Cel2Time and that caused genesis specs without gas limit or difficulty set to not have a default applied in Genesis.ToBlock

For this you can use `TestChainConfigNoCel2`, so that you don't need to touch those tests.

File: params/config.go [hunk 1]
```diff
@@ -269,7 +269,6 @@ var (
 		CancunTime:                    nil,
 		PragueTime:                    nil,
 		VerkleTime:                    nil,
-		Cel2Time:                      newUint64(0),
```
Comment (palango) [inline params/config.go]: Please revert this, we want to enable `Cel2` for as many tests as possible.

If a test fails please opt-out by using `TestChainConfigNoCel2`.

File: core/gen_genesis.go [hunk 2]
```diff
@@ -64,22 +64,22 @@ func (g Genesis) MarshalJSON() ([]byte, error) {
...
-		StateHash     *common.Hash                                `json:"stateHash,omitempty"`
+		Config        *params.ChainConfig                        `json:"config"`
+		Nonce         *math.HexOrDecimal64                       `json:"nonce"`
+		Timestamp     *math.HexOrDecimal64                       `json:"timestamp"`
+		ExtraData     *hexutil.Bytes                             `json:"extraData"`
+		GasLimit      *math.HexOrDecimal64                       `json:"gasLimit"`
+		Difficulty    *math.HexOrDecimal256                      `json:"difficulty"`
+		Mixhash       *common.Hash                               `json:"mixHash"`
+		Coinbase      *common.Address                            `json:"coinbase"`
+		Alloc         map[common.UnprefixedAddress]types.Account `json:"alloc"      gencodec:"required"`
```
Comment (palango) [inline core/gen_genesis.go:75]: Any idea why the type changes here? We should probably upstream that.

File: core/types/gen_header_rlp.go [hunk 3]
```diff
@@ -1,85 +0,0 @@
-// Code generated by rlpgen. DO NOT EDIT.
```
Comment (palango) [inline core/types/gen_header_rlp.go:1]: Why is this removed?

Comment (piersy) [reply core/types/gen_header_rlp.go:1]: Because we define our own EncodeRLP in `celo_block.go` that handles pre and post gingerbread rlp encodings.

Comment (palango) [reply core/types/gen_header_rlp.go:1]: Sure, but there it's just `rlp.Encode(w, &encodedHeader)` to encode it. If it's that easy, wthen why did all of this code exist? Will this have an impact on performance?

File: params/config.go (hunk 1 above)
Comment (piersy) [reply params/config.go]: The problem is that enabling Cel2 now causes many tests to fail:

Many tests do not specify gasLimit or difficulty in the geneses they create, assuming that those fields will be filled in by `Genesis.ToBlock` but if Cel2 is enabled they won't be filled in and then they fail. To fix all those tests we'd end up with a lot of changes scattered throughout the codebase.

It seems to me that the only tests that really need Cel2 enabled are our tests that are testing the Cel2 functionality, the rest can be run without Cel2 enabled.

File: core/types/gen_header_rlp.go (hunk 3 above)
Comment (piersy) [reply core/types/gen_header_rlp.go:1]: Yeah it was performance see here - https://github.com/ethereum-optimism/op-geth/commit/d6f49bf764c48e9415b168e701b6dc26566a675c

So we could probably add encoding like this back in for the 2 types we support. But I'd rather leave that for later since right now performance is not an issue.

File: core/gen_genesis.go (hunk 2 above)
Comment (piersy) [reply core/gen_genesis.go:75]: Hmm, no idea I ran `go generate` after removing required on gas limit and difficulty and this was the result.

Comment (piersy) [reply core/gen_genesis.go:75]: But GenesisAccount is an alias for types.Account so I don't think there should be any problem with this.

**Comment by piersy** [comment]
>> The decision on how to encode is made based on the value of gasLimit if zero then it's a pre-gingerbread header.

>I thought we would set that to the correct value, so that the apis return something useful. Or is this done somewhere in the API layer?

So we decided that we wouldn't add this to the encoded representation since it then breaks the assumption that `hash(rlp(header)) == header.Hash()`. In fact we strip out the `IstanbulAggregatedSeal` during migration so that `hash(rlp(header)) == header.Hash()`.

We can still return this at the api layer, but we are not going to put it in the rlp encoded header representation.

>> Additionally we needed to allow genesis blocks in a CeL2 context (by checking if Cel2Time is set) that lack GasLimit and Difficulty because those fields were considered required by op-geth but not present in early Celo.

>Is this requires so that the genesis hash stays the same, or why can't we just add those values to the celo genesis block?

It is precisely for the genesis hash, if those fields are set then the hash changes and then you can't connect to the network due to a genesis hash mismatch.

>> Many tests using TestChainConfig failed because it set Cel2Time and that caused genesis specs without gas limit or difficulty set to not have a default applied in Genesis.ToBlock

>For this you can use `TestChainConfigNoCel2`, so that you don't need to touch those tests.

Ok can have a look at how that works out.

File: core/types/gen_header_rlp.go (hunk 3 above)
Comment (piersy) [reply core/types/gen_header_rlp.go:1]: Added https://github.com/celo-org/op-geth/issues/145

File: core/types/celo_block.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	}
+
+	return err
+}
+
+// EncodeRLP implements encodes the Header to an RLP data stream.
+func (h *Header) EncodeRLP(w io.Writer) error {
+	// We check for a pre gingerbread header by looking for (GasLimit == 0)
+	// here. We don't use Difficulty  because CopyHeader can end up setting a
+	// nil Difficulty to a zero difficulty, so testing for nil difficulty is
```
Comment (alecps) [inline core/types/celo_block.go:90]: Why does setting a nil difficulty to a zero difficulty cause a problem? We are checking for a zero gaslimit, not a nil gaslimit, so I would think they would be the same right?

**Review by alecps — APPROVED** [review]

File: core/types/celo_block.go (hunk 4 above)
Comment (piersy) [reply core/types/celo_block.go:90]: Post gingerbread the difficulty is hardcoded to zero, so if we were to use `difficulty == 0` to select the encoding pre-gingerbread headers copied by `CopyHeader` would be encoded in a post gingerbread form.

File: params/config.go (hunk 1 above)
Comment (piersy) [reply params/config.go]: Ok I've raised a PR [here](https://github.com/celo-org/op-geth/pull/147) that reverts this change and makes use of `TestChainConfigNoCel2`.  It's not as much changes as I was envisaging but it's still quite a lot of extra scattered changes which will make updating from upstream more difficult. 

If the approach to fixing failed tests is simply to rely on `TestChainConfigNoCel2` then eventually (when all tests are using `TestChainConfigNoCel2`) we'll end up in the same situation as just not setting Cel2Time in the TestChainConfig but with a larger diff.

Also the approach of using `TestChainConfigNoCel2` is not particularly precise because sometimes we are setting it in a utility function and so it's effect is not confined to just the failing test.

So I think we would save ourselves some trouble if we didn't  add `Cel2Time` to the `TestConfig` and instead be aware of our specific changes and ensure that we test around them appropriately.

What do you think?

Just for reference, current changes we have that rely on `Cel2Time` are:

* Not setting gasLimit and difficulty to defaults in `genesis.ToBlock`
* Selecting transaction signers
* Setting exchange rates field in the block context
* Checking that the fee currency is whitelisted during the state transition
* Selecting the fee handler as the recipient of the base fee
* Adding celo precompiles to the set of precompiles acessible by the vm

File: core/gen_genesis.go (hunk 2 above)
Comment (palango) [reply core/gen_genesis.go:75]: This is already changes upstream, so is fine.

File: params/config.go (hunk 1 above)
Comment (palango) [reply params/config.go]: I still think it's worth having Cel2 flag enabled by default, so that we find potential errors when merging upstream.

I'd go with those changes that you did in the other branch. If we run into lot's of problems in the future during rebases, we can still change this.

Comment (piersy) [reply params/config.go]: Ok will merge that other branch in here.

Comment (piersy) [reply params/config.go]: Ok so actually this turned out to be a much larger task than it originally looked. Locally the go test caching failed me showing that tests were passing for https://github.com/celo-org/op-geth/pull/147, but actually they were not.

One of the problems with this is that the code panics which stops execution of the tests in a package, so you can't easily find all the failing tests, you have to fix one and then try again. Also now the diff is 4 times larger than #147 and tests are still failing, its not clear how much more will need to be changed.

So my feeling is that setting TestChainConfigNoCel2 on an individual test basis is not a sustainable approach.

@carterqw2 do you have any thoughts?

File: core/genesis.go [hunk 5]
```diff
@@ -475,11 +475,15 @@ func (g *Genesis) ToBlock() *types.Block {
...
 		Root:       root,
 	}
-	if g.GasLimit == 0 {
-		head.GasLimit = params.GenesisGasLimit
-	}
-	if g.Difficulty == nil && g.Mixhash == (common.Hash{}) {
-		head.Difficulty = params.GenesisDifficulty
+	// Disable setting default gas limit and difficulty in cel2 context.
+	// Because the celo genesis lacked gas limit and difficulty.
+	if g.Config.Cel2Time == nil {
```
Comment (carterqw2) [inline core/genesis.go]: Can we modify this check to something similar to the one [here](https://github.com/celo-org/op-geth/pull/143/files#diff-1272116434e5549f278a940767af52ed2d1abc91c44fa04ec48d8835a3774036R488) and check if Cel2 was activated in block 0 (which happens only in tests) and set values appropriately?

Comment (piersy) [reply core/genesis.go]: Yeah, I think this will work!

Comment (piersy) [reply core/genesis.go]: Thank you @carterqw2, that has solved the issue with all the tests!
